src/Exception/BatchException.php

```diff
@@ -0,0 +1,110 @@
+<?php
+
+namespace Http\Client\Exception;
+
+use Psr\Http\Message\ResponseInterface;
+
+/**
+ * @author GeLo <geloen.eric@gmail.com>
+ */
+final class BatchException extends TransferException
+{
+    /**
+     * @var TransferException[]
+     */
+    private $exceptions;
+
+    /**
+     * @var ResponseInterface[]
+     */
+    private $responses;
+
+    /**
+     * @param TransferException[] $exceptions
+     * @param ResponseInterface[]   $responses
+     */
+    public function __construct(array $exceptions = [], array $responses = [])
+    {
+        parent::__construct('An error occurred when sending multiple requests.');
+
+        foreach ($exceptions as $e) {
+            if (!$e instanceof TransferException) {
+                throw new InvalidArgumentException('Exception is not an instanceof Http\Client\Exception\TransferException');
+            }
+        }
+
+        foreach ($responses as $response) {
+            if (!$response instanceof ResponseInterface) {
+                throw new InvalidArgumentException('Response is not an instanceof Psr\Http\Message\ResponseInterface');
+            }
+        }
+
+
+        $this->exceptions = $exceptions;
+        $this->responses = $responses;
+    }
+
+    /**
+     * Returns all exceptions
+     *
+     * @return TransferException[]
+     */
+    public function getExceptions()
+    {
+        return $this->exceptions;
+    }
+
+    /**
+     * Checks if a specific exception exists
+     *
+     * @param TransferException $exception
+     *
+     * @return boolean TRUE if there is the exception else FALSE.
+     */
+    public function hasException(TransferException $exception)
+    {
+        return array_search($exception, $this->exceptions, true) !== false;
+    }
+
+    /**
+     * Checks if any exception exists
+     *
+     * @return boolean
+     */
+    public function hasExceptions()
+    {
+        return !empty($this->exceptions);
+    }
+
+    /**
+     * Returns all responses
```

Comment:
lets explain the semantics in the phpdoc to make it more useful and accessible for users. the doc says "Returns all responses" but if i understand you correctly, it should say "Returns all failed responses. The array keys are the same as in getExceptions". and we should also update the getExceptions phpdoc to say it has the same keys as getResponses. (though i expect most of those exceptions to carry the response in them again, no?). which brings me to wonder why we need this method at all? if there is an exception before we get a response, there is no response obviously. if we raise the exception because of a 4x or 5x response, we can put the response into that exception. what is the use case for having the responses separately?